book_maker/translator/chatgptapi_translator.py

```diff
@@ -90,5 +120,110 @@ def translate(self, text):
+        print(f"translation time: {elapsed_time:.1f}s")
+
         return t_text
+
+    def translate_and_split_lines(self, text):
+        result_str = self.translate(text, False)
+        lines = result_str.split("\n")
+        lines = [line.strip() for line in lines if line.strip() != ""]
+        return lines
+
+    def translate_list(self, plist):
+        sep = "\n\n\n\n\n"
+        # new_str = sep.join([item.text for item in plist])
+
+        new_str = ""
+        for p in plist:
+            temp_p = copy(p)
+            for sup in temp_p.find_all("sup"):
+                sup.extract()
+            new_str += temp_p.get_text().strip() + sep
+
+        if new_str.endswith(sep):
+            new_str = new_str[: -len(sep)]
+
+        plist_len = len(plist)
+
+        self.system_content = f"""{environ.get("OPENAI_API_SYS_MSG") or ""}
+
+Please translate the following paragraphs individually while preserving their original structure(This time it should be exactly {plist_len} paragraphs, no more or less).
+Only translate the paragraphs provided below:
+
+[Insert first paragraph here]
+
+[Insert second paragraph here]
+
+[Insert ... paragraph here]
+"""
+        # print(self.system_content)
+        print(f"plist len = {len(plist)}")
+
+        retry_count = 0
+        sleep_dur = 6
+        result_list = self.translate_and_split_lines(new_str)
+
+        start_time = time.time()
+        end_time = time.time()
+
+        best_result_list = result_list
+
+        while len(result_list) != plist_len and retry_count < 15:
+            print(
+                f"bug: {plist_len} -> {len(result_list)} : Number of paragraphs before and after translation"
+            )
+            print(f"sleep for {sleep_dur}s and retry {retry_count+1} ...")
+            time.sleep(sleep_dur)
+            result_list = self.translate_and_split_lines(new_str)
+            if (
+                len(result_list) == plist_len
+                or (
+                    len(result_list) > len(best_result_list)
+                    and len(result_list) <= plist_len
+                )
+                or (
+                    len(result_list) < len(best_result_list)
+                    and len(best_result_list) > plist_len
+                )
+            ):
+                best_result_list = result_list
+            retry_count += 1
+            end_time = time.time()
+
+        result_list = best_result_list
+
+        state = "fail" if len(result_list) != plist_len else "success"
+
+        if retry_count > 0:
+            print(f"retry {state}")
+            with open("buglog.txt", "a") as f:
+                print(
+                    f"retry {state}, count = {retry_count}, time = {(end_time-start_time):.1f}s",
+                    file=f,
+                )
+
+        if len(result_list) != plist_len:
+            # todo: select best
+            newlist = new_str.split(sep)
+            with open("buglog.txt", "a") as f:
```

Comment:
Create a `log` directory and add this directory to .gitignore.I think that would make the structure clearer:D